etmTk/etm.1
.\" Text automatically generated by txt2man
.TH etm 1 "06 January 2019" "version 3.2.38" "Unix user's manual"
.SH NAME
\fBetm \fP- manage events and tasks using simple text files
.SH SYNOPSIS
.nf
.fam C
  \fBetm\fP [\fIlogging\fP \fIlevel\fP] [\fIpath\fP] [?] [\fIacmsv\fP]
.fam T
.fi
.fam T
.fi
.SH DESCRIPTION
With no arguments, \fBetm\fP will use settings from the configuration file
~/.etm/etmtk.cfg, set \fIlogging\fP \fIlevel\fP 3 (warn) and open the GUI.
.PP
if the first argument is an integer not less than 1 (debug) and not
greater than 5 (critical), then \fBetm\fP will use that \fIlogging\fP \fIlevel\fP
and remove the argument.
.PP
If the first (remaining) argument is the \fIpath\fP to a directory which
contains a file named etm.cfg, then \fBetm\fP will use that configuration file
and remove the argument.
.PP
If the first (remaining) argument is one of the commands listed below,
then \fBetm\fP will execute the remaining arguments without opening the GUI.
.PP
.nf
.fam C
    a ARG   display the agenda view using ARG, if given, as a filter.
    k ARG   display the keywords view using ARG, if given, as a filter.
    n ARGS  Create a new item using the remaining arguments
            as the item specification.
    m INT   display a report using the remaining argument, which must be a
            positive integer, to display a report using the corresponding
            entry from the file given by report_specifications in etmtk.cfg.
            Use ? m to display the numbered list of entries from this file.
    p ARG   display the path view using ARG, if given, as a filter.
    r ARGS  display a report using the remaining arguments as the
            report specification.
    s ARG   display the schedule view using ARG, if given, as a filter.
    t ARG   display the tags view using ARG, if given, as a filter.
    v       display information about etm and the operating system.
    ? ARGS  display (this) command line help information if ARGS = '' or,
            if ARGS = X where X is one of the above commands, then display
            details about command X. 'X ?' is equivalent to '? X'.\

.fam T
.fi
.SH EXAMPLES
.SS    COMMAND LINE
Group items by year, month and day together
.PP
.nf
.fam C
    etm r c ddd, MMM d yyyy

.fam T
.fi
Output:
.PP
.nf
.fam C
    Fri, Apr 1 2011
      items for April 1
    Sat, Apr 2 2011
      items for April 2
    \.\.\.
.fam T
.fi
.SS    DATA FILES
Data items begin with a data type character and continue on one or more lines either until the end of the file is reached or another line is found that begins with a type character. Data type characters and the associated data types:
.TP
.B
\%~
Action: a record of time and/or money spent.
.TP
.B
\%*
Event: happens on a particular date and time.
.TP
.B
\%^
Occasion: happens on a particular date, e.g., a  holiday,
anniversary or birthday.
.TP
.B
\%!
Note: a record of some useful information.
.TP
.B
\%\-
Task: something that needs to be done.
.TP
.B
\%%
Delegated task: a task assigned to someone else.
.TP
.B
\%+
Task group: a group of related tasks, some of which may be
prerequisites for others.
.TP
.B
\%$
Inbasket: quick entry to be edited later when time permits.
.TP
.B
\%?
Someday maybe: remember but don't show in the common views.
.TP
.B
\%#
Hidden: remember but hide from all \fBetm\fP views except \fIpath\fP view.
.TP
.B
\%=
Defaults: set default options for subsequent entries in the
same data file.
.PP
The beginning data type character for each item is followed by the item summary and then, perhaps, by one or more '@key value' option pairs. Examples:
.IP \(bu 3
A sales meeting (an event) a week from today from 9:00am until 10:00am with a 5 minute early warning alert:
.PP
.nf
.fam C
    \%* sales meeting @s +7 9a @e 1h @a 5

.fam T
.fi
.IP \(bu 3
Prepare a report (a task) for the meeting beginning 3 days early:
.PP
.nf
.fam C
    \%\- prepare report @s +7 @b 3

.fam T
.fi
.IP \(bu 3
A 35 minute period (an action) spent working on the report yesterday:
.PP
.nf
.fam C
    \%~ report preparation @s \-1 @e 35

.fam T
.fi
.IP \(bu 3
Get a haircut (a task) on the 24th of the current month and then [r]epeatedly at (d)aily [i]ntervals of 14 days and, [o]n completion, (r)estart from the completion date:
.PP
.nf
.fam C
    \%\- get haircut @s 24 @r d &i 14 @o r

.fam T
.fi
.IP \(bu 3
Do the jobs in the following task group in 'q' order to finish the dog house project:
.PP
.nf
.fam C
    \%+ dog house
          @j pickup lumber and paint &q 1
          @j cut pieces &q 2
          @j assemble &q 3
          @j paint &q 4

.fam T
.fi
.IP \(bu 3
Payday (an occassion) on the last week day of each month. The '&s' part of the entry extracts the last date which is both a weekday and falls within the last three days of the month.):
.PP
.nf
.fam C
    \%^ payday @s 1/1 @r m &w (MO, TU, WE, TH, FR) &m (\-1, \-2, \-3) &s \-1

.fam T
.fi
.IP \(bu 3
Take a prescribed medication daily (a reminder) for the next three days at 10am, 2pm, 6pm and 10pm and trigger the default alert zero minutes before each event:
.PP
.nf
.fam C
    \%* take Rx @s +0 @r d &h 10, 14, 18, 22 &u +4 @a 0

.fam T
.fi
.IP \(bu 3
Presidential election day (an occassion) every four years on the first Tuesday after a Monday in November:
.PP
.nf
.fam C
    \%^ Presidential Election Day @s 2012-11-06
          @r y &i 4 &M 11 &m range(2,9) &w TU

.fam T
.fi
.IP \(bu 3
Join the \fBetm\fP discussion group (a task). Because of the @g (goto) link, pressing Ctrl-G when the details of this item are displayed in the gui would open the link using the system default application:
.PP
.nf
.fam C
    \%\- join the etm discussion group
             @g http://groups.google.com/group/eventandtaskmanager/topics
.fam T
.fi
.SH SEE ALSO
Extensive documentation can be found in the folder:
.PP
.nf
.fam C
    http://people.duke.edu/~dgraham/etmtk/help/
.fam T
.fi
.SH BUGS
Please report bugs to the \fBetm\fP discussion group:
.PP
.nf
.fam C
    http://groups.google.com/forum/#!forum/eventandtaskmanager
.fam T
.fi
.SH AUTHOR
Daniel A Graham <daniel.graham@duke.edu>
.SH COPYRIGHT
Copyright (c) 2009-2014 [Daniel Graham]. All rights reserved.


CHANGES
# Changes in the 2 weeks preceding Fri Jun 13 06:30:32 EDT 2014:
- 2014-06-13 06:31:06 -0400 (HEAD, tag: 3.0.11, master): Dan Graham
   tagged version 3.0.11
- 2014-06-13 06:24:32 -0400 (origin/master): Dan Graham
   Tweaks for mk_docs and upload.
- 2014-06-13 06:12:46 -0400: Dan Graham
   Merge branch 'master' of https://github.com/dagraham/etm-tk
- 2014-06-12 22:28:54 -0400 (tag: 3.0.13): Dan Graham
   tagged version 3.0.13
- 2014-06-12 22:28:41 -0400: Dan Graham
   Version number catchup.
- 2014-06-12 22:13:48 -0400: Dan Graham
   moved help files to etm-tk.wiki.
- 2014-06-12 17:43:23 -0400 (tag: 3.0.12): Dan Graham
   tagged version 3.0.12
- 2014-06-12 16:28:49 -0400 (tag: 3.0.9-7): Dan Graham
   tagged version 3.0.11
- 2014-06-12 16:24:46 -0400: Dan Graham
   Keep UserManual pdf and html.
- 2014-06-12 15:55:01 -0400: Dan Graham
   Updated version numbers and added INDEX.md.
- 2014-06-12 15:52:17 -0400: Dan Graham
   Documentation reorganization.
- 2014-06-12 10:14:39 -0400 (tag: 3.0.10): Dan Graham
   tagged version 3.0.10
- 2014-06-12 10:14:26 -0400: Dan Graham
   Added tkinter >= 8.5.11 to REQUIRES.
- 2014-06-12 10:02:25 -0400: Dan Graham
   Removed Windows XP from the trove list of supported platforms.
- 2014-06-12 08:14:13 -0400: Dan Graham
   Documentation updates.
- 2014-06-11 16:16:39 -0400 (tag: 3.0.9): Dan Graham
   tagged version 3.0.9
- 2014-06-11 16:02:36 -0400: Dan Graham
   Added warning that timezone information is ignored to result from
   date calculator for pre 1970 dates under Windows.
- 2014-06-11 13:00:43 -0400: Dan Graham
   Eliminated use of the .etmtkdata.pkl file and, with it, the need to
   import pickle and the functions dump_data and load_data.  Fixed bug
   in date calculator regex. Restored ability to use "easter(yyyy)" in
   date calculator.
- 2014-06-10 16:55:19 -0400 (tag: 3.0.8): Dan Graham
   tagged version 3.0.8
- 2014-06-10 15:24:52 -0400: Dan Graham
   Added hack to enable Windoz to deal with pre 1970-01-01 datetimes. 
   Corrected argument call to get_changes.
- 2014-06-10 11:44:10 -0400 (tag: 3.0.7): Dan Graham
   tagged version 3.0.7
- 2014-06-10 11:40:02 -0400: Dan Graham
   Only use logfile on startup if etmdir exists.  Dump and load
   uuid2labels in pickle file.  Only remove pickle file with bad data
   if it exists. Added &E, byeaster, to rrule options.
- 2014-06-09 13:03:11 -0400: Dan Graham
   Added Shift-Ctrl-F shortcut for clear filter.
- 2014-06-09 10:12:35 -0400: Dan Graham
   Added missing nullok to opts for startActionTimer.
- 2014-06-09 07:06:24 -0400 (tag: 3.0.6): Dan Graham
   tagged version 3.0.6
- 2014-06-09 07:05:22 -0400: Dan Graham
   Initialize edithsh to avoid error when editing files.
- 2014-06-08 12:58:42 -0400: Dan Graham
   Fixed bug in reference to self.options instead of loop.options in
   edit.py.
- 2014-06-08 07:38:43 -0400: Dan Graham
   Delete content on switch to either week or month view.
- 2014-06-07 07:22:09 -0400 (tag: 3.0.5): Dan Graham
   tagged version 3.0.5
- 2014-06-07 07:21:55 -0400: Dan Graham
   Use isocalendar week numbers in month view (as in week view).
   Updated screenshots.
- 2014-06-07 06:39:47 -0400 (tag: 3.0.4): Dan Graham
   tagged version 3.0.4
- 2014-06-06 17:36:08 -0400: Dan Graham
   Make sure tree has focus on startup, even if there are no data
   items.
- 2014-06-06 15:43:35 -0400: Dan Graham
   Remove confirmation for new event on double-click in week view.
- 2014-06-06 15:28:05 -0400: Dan Graham
   Don't insert ? and @e 1h when creating a new event in week view.
- 2014-06-06 15:17:24 -0400: Dan Graham
   When pressing "N" to create a new item in day, week or month view,
   start with @s set to the currently selected date. Also for Return
   or Double-Click in month view.
- 2014-06-06 11:40:07 -0400 (tag: 3.0.3): Dan Graham
   tagged version 3.0.3
- 2014-06-06 11:39:29 -0400: Dan Graham
   Fixed bug in closeMonthly.
- 2014-06-06 10:53:47 -0400: Dan Graham
   Create etmtk_log.txt if it does not exist in get_options.
- 2014-06-06 10:41:31 -0400: Dan Graham
   Added busy and free times lists to month view.
- 2014-06-05 16:29:15 -0400: Dan Graham
   Updated images and README.
- 2014-06-05 13:06:42 -0400 (tag: 3.0.1): Dan Graham
   tagged version 3.0.1
- 2014-06-05 13:05:46 -0400: Dan Graham
   Added monthly screenshot.
- 2014-06-05 13:05:22 -0400: Dan Graham
   Respect active calendars in month view. Updated shortcuts. Added
   Month view to overview.
- 2014-06-05 12:03:37 -0400: Dan Graham
   Month view works with up, down, left, right and space bindings. Day
   number color reflects number of minutes scheduled.
- 2014-06-03 18:56:54 -0400: Dan Graham
   Background color tweaks. Hide insertion cursor in ReadOnlyText.
- 2014-06-03 06:52:27 -0400: Dan Graham
   Merge branch 'master' of https://github.com/dagraham/etm-tk
- 2014-06-03 06:50:31 -0400 (tag: 3.0.0): Dan Graham
   tagged version 3.0.0
- 2014-06-03 06:49:06 -0400: Dan Graham
   Restart version numbering at 3.0.0.
- 2014-06-03 06:43:05 -0400: Dan Graham
   Changed idle_minum to idle_minutes.  Changed development status to
   5 - Production/Stable.
- 2014-06-02 22:50:06 -0400: Dan Graham
   Merge branch 'master' of https://github.com/dagraham/etm-tk
- 2014-06-02 22:48:09 -0400 (tag: 0.2.1): Dan Graham
   tagged version 0.2.1
- 2014-06-02 22:47:34 -0400: Dan Graham
   Changed development status to 5 - Production/Stable and version to
   0.2.0.
- 2014-06-02 22:31:44 -0400: Dan Graham
   tagged version 0.2.1
- 2014-06-02 22:31:12 -0400: Dan Graham
   Changed development status to 5 - Production/Stable and version to
   0.2.0.
- 2014-06-02 14:59:46 -0400 (tag: 0.1.44): Dan Graham
   tagged version 0.1.44
- 2014-06-02 14:59:35 -0400: Dan Graham
   Fixed bug in passing font to DialogWindow.
- 2014-06-02 13:15:10 -0400 (tag: 0.1.43): Dan Graham
   tagged version 0.1.43
- 2014-06-02 13:14:46 -0400: Dan Graham
   Added available dates screenshots.
- 2014-06-02 13:13:53 -0400: Dan Graham

etm.py
# -*- coding: utf-8 -*-
import sys
import etmKv.data as data
from etmKv.data import get_current_time, init_localization, str2hsh, hsh2str

from kivy.app import App
from kivy.uix.boxlayout import BoxLayout
from kivy.uix.textinput import TextInput
from kivy.uix.codeinput import CodeInput
from kivy.uix.button import Button
# from kivy.uix.widget import Widget

from pygments.lexers.special import TextLexer

from kivy.uix.label import Label
from kivy.uix.modalview import ModalView
from kivy.properties import ObjectProperty
from kivy.config import Config
from kivy.clock import Clock
Config.set('graphics', 'height', '420')
Config.set('graphics', 'width', '520')

import gettext
_ = gettext.gettext


class ETMEditor(CodeInput):

    def __init__(self, parent=None, callback=None):
        self.parent = parent
        self.button = Button
        self.callback = callback
        super(ETMEditor, self).__init__(multiline=True, size_hint=(1, None), height=300, lexer=TextLexer())
        self.modified = False
        self.bind(focus=self.on_focus)

    def _keyboard_on_key_up(self, window, keycode):
        # inform ETMDialog about the modification state
        if self.callback:
            self.callback((len(self._undo) > 0))
        super(ETMEditor, self)._keyboard_on_key_up(
            window, keycode)

    def on_focus(self, instance, value):
        if value:
            print('Editor focused', instance)
        else:
            print('Editor defocused', instance)
        super(ETMEditor, self).on_focus(instance, value)


class ETMDialog():

    def __init__(self, parent=None, options={}):
        self.parent = parent
        self.options = options
        self.editor = ETMEditor()
        self.text = ''

        self.btnvalidate = Button(text=_('Close'), size_hint_y=None, height='30sp')
        self.btncancel = Button(text=_('Cancel'), size_hint_y=None, height='30sp')
        buttons = BoxLayout(orientation='horizontal', height='30sp')
        buttons.add_widget(self.btncancel)
        buttons.add_widget(self.btnvalidate)
        self.message = Label(text="", size_hint_y=None, height=80)
        self.content = BoxLayout(orientation='vertical')

        self.input = ETMEditor(callback=self.set_modified)
        self.input.background_color = [1, 1, 1, 1]
        self.input.font_size =  self.options['fontsize']

        # self.input.bind(on_text_validate=self.validate)
        # self.input.bind(on_key_up=self.on_text)
        self.content.add_widget(self.input)
        self.content.add_widget(buttons)
        self.content.add_widget(self.message)
        self.popup = ModalView(size_hint=(None, None), size=(500, 420))
        self.popup.add_widget(self.content)
        self.btnvalidate.bind(on_release=self.validate)
        self.btncancel.bind(on_release=self.popup.dismiss)

    def run(self, text="", msg=""):
        print('run', text, msg)
        self.input.text = text
        self.message.text = msg
        self.input.reset_undo()
        self.input.focus = True
        self.popup.open()

    def validate(self, value):
        new_str = ''
        if self.modified:
            new_hsh, msg = str2hsh(self.input.text, options=self.options)
            if msg:
                self.message.text = "\n".join(msg)
            else:
                try:
                    new_str = hsh2str(new_hsh, self.options)
                    new_item = unicode(u"{0}".format(hsh2str(new_hsh, self.options)))
                    self.input.text = new_item
                    self.message.text = self.parent.loop.do_n(arg_str=new_item)
                    self.set_modified(False)
                except:
                    self.message.text = "Error processing %s" % new_hsh
        else:
            if new_str:
                self.parent.output_wid.text = new_str
            self.popup.dismiss()


    def set_modified(self, bool):
        print('set_modified', bool)
        self.modified = bool
        if bool:
            self.btnvalidate.text = _('Validate')
        else:
            self.btnvalidate.text = _('Close')


class ETMTextInput(TextInput):

    history = []
    index = 0
    now = None
    options = {}
    popup = ''
    value = ''
    firsttime = True
    mode = 'command'   # or edit or delete
    item_hsh = {}

    def _keyboard_on_key_down(self, window, keycode, text, modifiers):
        # print(keycode)
        if keycode[0] == 273:  # 273 is the keycode for up
            self.previous_history()
        elif keycode[0] == 274:  # 274 is the keycode for down
            self.next_history()
        elif keycode[0] == 13:  # 13 is the keycode for enter
            self.process_input()
        elif keycode[0] == 27:  # escape
            self.select_all()
            self.delete_selection()
            # return True to avoid closing app
            return(True)
        else:
            super(ETMTextInput, self)._keyboard_on_key_down(
                window, keycode, text, modifiers)

    def init(self):
        if self.firsttime:
            # only do this once
            self.firsttime = False
            self.bind(focus=self.on_focus)

            self.loop = loop
            self.options = self.loop.options
            self.Dialog = ETMDialog(parent=self, options=self.options)
            self.loop.parent = self
            res = self.loop.do_command('a')
            self.start_timer()
            return(res)
        return()

    def start_timer(self):
        self.now = get_current_time()
        nxt = (60 - self.now.second)
        print(self.now)
        Clock.schedule_once(self.timer_callback, nxt)

    def timer_callback(self, dt):
        self.start_timer()

    def get_input(self, question):
        self.command_mode = False
        return(question)

    def process_input(self):
        """
        This is called whenever enter is pressed in the input field.
        Action depends upon comand_mode.
        Append input to history, process it and show the result in output.
        """
        cmd = self.text.strip()

        if not cmd:
            return(True)

        if self.mode == 'command':
            if cmd == 'q':
                sys.exit()
            cmd = cmd.strip()
            if cmd[0] in ['a', 'r', 't']:
                # simple command history for report commands
                if cmd in self.history:
                    self.history.remove(cmd)
                self.history.append(cmd)
                self.index = len(self.history) - 1
            # select everything in input to make it easy to clear
            self.select_all()
            try:
                res = loop.do_command(cmd)
            except:
                return _('could not process command "{0}"').format(cmd)

                # self.Dialog = ETMDialog(parent=self, text='except')
                self.Dialog.run()
                res = self.Dialog.text
                print('res', res)

        elif self.mode == 'edit':
            print('edit', cmd)
            res = loop._do_edit(cmd)

        elif self.mode == 'delete':
            print('deleted', cmd)
            res = ''

        elif self.mode == 'new_date':
            print('date', cmd)
            res = loop._new_date(cmd)

        if not res:
            return(True)
        print('display and scroll to top')
        self.output_wid.text = res
        self.scroll_wid.scroll_y = 1
        self.scroll_to_top()

    def scroll_to_top(self):
        # print('scrolling')
        self.scroll_wid.update_from_scroll()

    def previous_history(self):
        """
        Replace input with the previous history item.
        """
        if self.index >= 1:
            self.index -= 1
            self.text = self.history[self.index]

    def next_history(self):
        """
        Replace input with the next history item.
        """
        if self.index + 1 < len(self.history):
            self.index += 1
            self.text = self.history[self.index]

    def on_focus(self, instance, value):
        if value:
            print('focused', instance)
        else:
            print('defocused', instance)
        super(ETMTextInput, self).on_focus(instance, value)


class BoxIOWidget(BoxLayout):
    input_wid = ObjectProperty()
    status_wid = ObjectProperty()
    output_wid = ObjectProperty()
    scroll_wid = ObjectProperty()


class etmApp(App):
    title = 'etm'
    icon = 'etmlogo_512x512x32.png'

    def build(self):
        return BoxIOWidget()

if __name__ == "__main__":
    init_localization()
    etmdir = ''
    if len(sys.argv) > 1:
        etmdir = sys.argv.pop(1)
    (user_options, options, use_locale) = data.get_options(etmdir)
    loop = data.ETMCmd(options)
    etmApp().run()
